automod/fetch.go

```diff
@@ -0,0 +1,113 @@
+	"github.com/bluesky-social/indigo/atproto/identity"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/bluesky-social/indigo/xrpc"
+)
+
+func (e *Engine) FetchAndProcessRecord(ctx context.Context, aturi syntax.ATURI) error {
+	// resolve URI, identity, and record
+	if aturi.RecordKey() == "" {
+		return fmt.Errorf("need a full, not partial, AT-URI: %s", aturi)
+	}
+	ident, err := e.Directory.Lookup(ctx, aturi.Authority())
+	if err != nil {
+		return fmt.Errorf("resolving AT-URI authority: %v", err)
+	}
+	pdsURL := ident.PDSEndpoint()
+	if pdsURL == "" {
+		return fmt.Errorf("could not resolve PDS endpoint for AT-URI account: %s", ident.DID.String())
+	}
+	pdsClient := xrpc.Client{Host: ident.PDSEndpoint()}
+
+	e.Logger.Info("fetching record", "did", ident.DID.String(), "collection", aturi.Collection().String(), "rkey", aturi.RecordKey().String())
+	out, err := comatproto.RepoGetRecord(ctx, &pdsClient, "", aturi.Collection().String(), ident.DID.String(), aturi.RecordKey().String())
+	if err != nil {
+		return fmt.Errorf("fetching record from Relay (%s): %v", aturi, err)
+	}
+	if out.Cid == nil {
+		return fmt.Errorf("expected a CID in getRecord response")
+	}
+	return e.ProcessRecord(ctx, ident.DID, aturi.Path(), *out.Cid, out.Value.Val)
+}
+
+func (e *Engine) FetchRecent(ctx context.Context, atid syntax.AtIdentifier, limit int) (*identity.Identity, []*comatproto.RepoListRecords_Record, error) {
+	ident, err := e.Directory.Lookup(ctx, atid)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to resolve AT identifier: %v", err)
+	}
+	pdsURL := ident.PDSEndpoint()
+	if pdsURL == "" {
+		return nil, nil, fmt.Errorf("could not resolve PDS endpoint for account: %s", ident.DID.String())
+	}
+	pdsClient := xrpc.Client{Host: ident.PDSEndpoint()}
+
+	resp, err := comatproto.RepoListRecords(ctx, &pdsClient, "app.bsky.feed.post", "", int64(limit), ident.DID.String(), false, "", "")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch record list: %v", err)
+	}
+	e.Logger.Info("got recent posts", "did", ident.DID.String(), "pds", pdsURL, "count", len(resp.Records))
+	return ident, resp.Records, nil
+}
+
+func (e *Engine) FetchAndProcessRecent(ctx context.Context, atid syntax.AtIdentifier, limit int) error {
+
+	ident, records, err := e.FetchRecent(ctx, atid, limit)
+	if err != nil {
+		return err
+	}
+	// records are most-recent first; we want recent but oldest-first, so iterate backwards
+	for i := range records {
+		rec := records[len(records)-i-1]
+		aturi, err := syntax.ParseATURI(rec.Uri)
+		if err != nil {
+			return fmt.Errorf("parsing PDS record response: %v", err)
+		}
+		err = e.ProcessRecord(ctx, ident.DID, aturi.Path(), rec.Cid, rec.Value.Val)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type AccountCapture struct {
+	CapturedAt  syntax.Datetime                     `json:"capturedAt"`
+	AccountMeta AccountMeta                         `json:"accountMeta"`
```

Comment:
Okay, main question: if there will be rules that want to look at multiple identities, should we brace for that now by making this a list or a map?  Or is that overkill.

Reply:
That is a reasonable question, but I think it is less ergonomic and maybe overkill. For simple rules it is nice to have a direct struct field to get access to the account meta for the owner/creator/author of the repo/record, which is by far the common access case.

For fetching more account meta, I think the pattern we should use is tiered caching. The rule code should just call `evt.GetAccountMeta(did-or-handle)`, and the engine should read-through whatever layers of cache to get that info. That caching may even include per-event caching (not implemented currently!) which would be 1) fast and 2) ensure consistent behavior between rules executed on the same event (aka, the whole event should see the same account meta for all accounts).

This is following the facebook FXL functional pattern of basically memoizing all function calls within the scope of an event.

(side note: still mulling a better name than "event")

Reply:
:thought_balloon: I guess we could have an internal `accountMetaMap` and have a helper method (`CurrentAccount()`) which would access the current account?